rustls/src/webpki/verify.rs

```diff
@@ -408,42 +414,135 @@ impl From<webpki::Error> for CertRevocationListError {
     }
 }
 
-static ECDSA_SHA256: SignatureAlgorithms = &[webpki::ECDSA_P256_SHA256, webpki::ECDSA_P384_SHA256];
-
-static ECDSA_SHA384: SignatureAlgorithms = &[webpki::ECDSA_P256_SHA384, webpki::ECDSA_P384_SHA384];
-
-static ED25519: SignatureAlgorithms = &[webpki::ED25519];
-
-static RSA_SHA256: SignatureAlgorithms = &[webpki::RSA_PKCS1_2048_8192_SHA256];
-static RSA_SHA384: SignatureAlgorithms = &[webpki::RSA_PKCS1_2048_8192_SHA384];
-static RSA_SHA512: SignatureAlgorithms = &[webpki::RSA_PKCS1_2048_8192_SHA512];
-static RSA_PSS_SHA256: SignatureAlgorithms = &[webpki::RSA_PSS_2048_8192_SHA256_LEGACY_KEY];
-static RSA_PSS_SHA384: SignatureAlgorithms = &[webpki::RSA_PSS_2048_8192_SHA384_LEGACY_KEY];
-static RSA_PSS_SHA512: SignatureAlgorithms = &[webpki::RSA_PSS_2048_8192_SHA512_LEGACY_KEY];
+/// Describes which `webpki` signature verification algorithms are supported and
+/// how they map to TLS `SignatureScheme`s.
+#[derive(Clone, Copy)]
+#[allow(unreachable_pub)]
+pub struct WebPkiSupportedAlgorithms {
+    /// A list of all supported signature verification algorithms.
+    ///
+    /// Used for verifying certificate chains.
+    ///
+    /// The order of this list is not significant.
+    pub all: &'static [&'static dyn SignatureVerificationAlgorithm],
 
-fn convert_scheme(scheme: SignatureScheme) -> Result<SignatureAlgorithms, Error> {
-    match scheme {
-        // nb. for TLS1.2 the curve is not fixed by SignatureScheme.
-        SignatureScheme::ECDSA_NISTP256_SHA256 => Ok(ECDSA_SHA256),
-        SignatureScheme::ECDSA_NISTP384_SHA384 => Ok(ECDSA_SHA384),
+    /// A mapping from TLS `SignatureScheme`s to matching webpki signature verification algorithms.
+    ///
+    /// This is one (`SignatureScheme`) to many ([`SignatureVerificationAlgorithm`]) because
+    /// (depending on the protocol version) there is not necessary a 1-to-1 mapping.
+    ///
+    /// For TLS1.2, all `SignatureVerificationAlgorithm`s are tried in sequence.
+    ///
+    /// For TLS1.3, only the first is tried.
+    ///
+    /// The supported schemes in this mapping is communicated to the peer and the order is significant.
+    /// The first mapping is our highest preference.
+    pub mapping: &'static [(
```

Comment:
Why don't we want to this to just be a top-level `static`?

Reply:
not quite sure what you mean. it needs to be configurable rather than static so the `WebPkiServerVerifier` (etc.) can be reused without ring. there's a separate constructor of this type in https://github.com/rustls/rustls/pull/1405/commits/ba26a6d7c5037fb30362640ed33b43cdff5ace22#diff-8749c6ccb1619fb20bb8abc0a3651cc58eaf05ff79a0c395c7b9ef76047cea05R11-R17 backed by https://crates.io/crates/rsa if that makes the goal any clearer?

Reply:
Okay, so this is really a way for the crypto provider to provide it's signature verification algorithms, and a mapping from `SignatureScheme` to them? And you used the `WebPki` prefix since this is specifically used for the webpki-backed verifiers?

Reply:
I don't know, it feels more like a crypto provider concern that happens to be used by the verifiers, not all that webpki-specific.

Reply:
I think I'm convinced that this a decent design. Should we be worried about the link-time inclusion of all signature verification algorithms by default?

Reply:
I think that is the status quo. Though I guess this does now give people the opportunity to avoid that if desired.